cmd/build/v2/services.go

```diff
@@ -27,45 +27,43 @@ import (
 )
 
 // GetServicesToBuild returns the services it has to built because they are not already built
-func (bc *OktetoBuilder) GetServicesToBuild(ctx context.Context, manifest *model.Manifest, svcToDeploy []string) ([]string, error) {
+func (bc *OktetoBuilder) GetServicesToBuild(ctx context.Context, manifest *model.Manifest, servicesToDeploy []string) ([]string, error) {
 	buildManifest := manifest.Build
 
 	// check if images are at registry (global or dev) and set envs or send to build
-	toBuild := make(chan string, len(buildManifest))
+	nonBuiltServices := make(chan string, len(buildManifest))
 	g, _ := errgroup.WithContext(ctx)
 	for service := range buildManifest {
 
 		svc := service
 		g.Go(func() error {
-			return bc.checkServicesToBuild(svc, manifest, toBuild)
+			return bc.checkServicesToBuild(svc, manifest, nonBuiltServices)
 		})
 	}
 
 	if err := g.Wait(); err != nil {
 		return nil, err
 	}
-	close(toBuild)
+	close(nonBuiltServices)
 
-	if len(toBuild) == 0 {
+	if len(nonBuiltServices) == 0 {
 		oktetoLog.Information("Images were already built. To rebuild your images run 'okteto build' or 'okteto deploy --build'")
 		if err := manifest.ExpandEnvVars(); err != nil {
 			return nil, err
 		}
 		return nil, nil
 	}
 
-	svcToDeployMap := map[string]bool{}
-	for _, svc := range svcToDeploy {
-		svcToDeployMap[svc] = true
-	}
-	svcsToBuildList := []string{}
-	for svc := range toBuild {
-		if _, ok := svcToDeployMap[svc]; len(svcToDeploy) > 0 && !ok {
-			continue
-		}
-		svcsToBuildList = append(svcsToBuildList, svc)
+	servicesToDeploySet := sliceToSet(servicesToDeploy)
+	nonBuiltServicesSet := channelToSet(nonBuiltServices)
+
+	servicesToBuildSet := setIntersection(nonBuiltServicesSet, servicesToDeploySet)
+
+	if len(servicesToBuildSet) == 0 {
+		return setToSlice(nonBuiltServicesSet), nil
+	} else {
+		return setToSlice(servicesToBuildSet), nil
```

Comment:
Agree. I've also seen this weird behavior.

Based on your comments, seems like we'd need the union rather than the intersection.

I'll change the tests and then this to be the union.

Reply:
I've changed this and I tried to follow the naming as close as it was before. I want to focus more on the logic around this to reduce the probability of introducing bugs.